### PR TITLE
feat: Return a hook from the createContainer method

### LINKED ISSUE #77: Suggestion for API Tweak
What do you think of this API? It's a very simple change to the library, but personally I find it more elegant.

```javascript
import React, { useState, useContext } from "react";
import createContainer from "constate";

// 1️⃣ Create a custom hook as usual
const useCounter = () => {
  const [count, setCount] = useState(0);
  const increment = () => setCount(prevCount => prevCount + 1);
  return { count, increment };
}

// 2️⃣ When you need to share your state, simply wrap your hook
//    with the createContainer higher-order hook, like so:
const useCounter = createContainer(() => {
   // ... same logic here
});

function Button() {
  // 3️⃣ Use the same hook like you were before, with the same api. No change.
  const { increment } = useCounter()
  return <button onClick={increment}>+</button>;
}

function Count() {
  // 4️⃣ But now you can use it in other components as well.
  const { count } = useCounter()
  return <span>{count}</span>;
}

function App() {
  // 5️⃣ The caveat, you wrap your components with the Provider that is
  //    attached to the hook
  return (
    <useCounter.Provider>
      <Count />
      <Button />
    </useCounter.Provider>
  );
}
```

The change in the code is roughly this (the last 3 lines specifically):

```javascript
export function createContainer(useValue, createMemoInputs) {
  const Context = React.createContext()

  const Provider = props => {
    const value = useValue(props)
    const { children } = props
    const memoizedValue = createMemoInputs ? React.useMemo(() => value, createMemoInputs(value)) : value
    return <Context.Provider value={memoizedValue}>{children}</Context.Provider>
  }

  const hook = () => useContext(Context)
  hook.Provider = Provider
  return hook
}
```

### PR DESCRIPTION
**Pull request by timkindberg**
- Update README.md examples to showcase this as the preferred method (not sure if this is to presumptuous)

Fixes #77

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,36 +39,38 @@ import React, { useState, useContext } from "react";
...
   const [count, setCount] = useState(0);
   const increment = () => setCount(prevCount => prevCount + 1);
   return { count, increment };
 }
 
-// 2️⃣ Create container
-const CounterContainer = createContainer(useCounter);
+// 2️⃣ When you need to share your state, simply wrap your hook
+//    with the createContainer higher-order hook, like so:
+const useCounter = createContainer(() => {
```
Comment (diegohaz) [inline README.md]: Let's call it
```
const useCounterContainer = createContainer(useCounter);
```

File: README.md [hunk 2]
```diff
@@ -39,36 +39,38 @@ import React, { useState, useContext } from "react";
...
+   // ... same logic here
+});
 
 function Button() {
-  // 3️⃣ Use container context instead of custom hook
-  // const { increment } = useCounter();
-  const { increment } = useContext(CounterContainer.Context);
+  // 3️⃣ Use the same hook like you were before, with the same api. No change.
+  const { increment } = useCounter()
```
Comment (diegohaz) [inline README.md]: The hook returned by `createContainer` isn't 100% equivalent to the wrapped hook. You can't pass arguments to the hook, for example.

File: src/index.tsx [hunk 3]
```diff
@@ -34,10 +34,10 @@ function createContainer<P, V>(
...
   };
 
-  return {
-    Context,
-    Provider
-  };
+  const hook = () => React.useContext(Context);
+  hook.Context = Context;
+  hook.Provider = Provider;
+  return hook;
```
Comment (diegohaz) [inline src/index.tsx]: Let's call it `useContainer` instead of `hook`.

**Comment by diegohaz** [comment]
I'm also thinking about these names:
```jsx
import createUseContext from "constate";

const useCounterContext = createUseContext(useCounter);
```

It wouldn't affect current users since it's a default export.

**Comment by diegohaz** [comment]
@bySabi What do you think about this change?

**Comment by bySabi** [comment]
The changes looks good, although I completely removed all references to the Context, they make the constate API tight to its implementation.

I would add something like:
 
```
import createHookContainer from "constate";

const useCounterCtx = createHookContainer(useCounter);
```

or

```
import createHook from "constate";

const useCounterCtx = createHook(useCounter);
```

I'm not sure too it can be called either **higher-order hook** to `createContainer`

**Comment by timkindberg** [comment]
@bySabi I think using the word Context might be necessary because we are still requiring devs to set up the Provider, and devs know that a Context is the only way to work with Provider values. So in this case the implementation should probably stay transparent to the dev. I like the word "Context" better than "Container", because it's an existing word in the React vocabulary, whereas "Container" is not.

**Comment by bySabi** [comment]
a `Container` in any XML based document refer to any enclosing tags that contains a sub-tree of nodes, on the other hand React Context is pattern that can be use a **Container** like `Context.Provider` for define a scope but is not require when fallback default is use.
In the future `constate` could be reimplemented without Context,  for performance reasons mainly just as is doing Redux 7 and the semantics of its API would not be tight to Context.

**Comment by timkindberg** [comment]
> In the future `constate` could be reimplemented without Context, for performance reasons mainly just as is doing Redux 7 and the semantics of its API would not be tight to Context.

^ That will be great. 

However, for now the implementation is still leaking out; it's obvious we are using Provider/Context to the user. So until we can "hide" it completely, I personally feel that being open and clear with the dev is a better API. With that said, it's Diego's lib, so he can name it however he feels. That's just our two cents each :)

@diegohaz I submitted my pull request based of your naming suggestions but focused more on the word "Context". I found the examples were yelling at me because they only work with a published version of the lib (1.0.0 in this case). So I'm not sure what do to there...

I just pushed the updates, LMK what you think...

**Comment by diegohaz** [comment]
> With that said, it's Diego's lib, so he can name it however he feels.

I'm very bad at this, specially because English is not my native language. 😅 I try my best, but feedback is really appreciated.

About the PR, it looks good to me. I'd make some small changes in the readme, but I can do that later.

**Comment by timkindberg** [comment]
Thank you @diegohaz! Thank you for being open to the change.

**Pull request merged**